specs/AdditionalAllowedFrameAncestors.md

```diff
@@ -0,0 +1,234 @@
+Additional Allowed Frame Ancestors for iframes
+===
+
+# Background
+Due to potential  [Clickjacking](https://en.wikipedia.org/wiki/Clickjacking) attack, a lot of sites only allow themselves to be hosted in certain trusted ancestor iframes and pages. The main way to specify this ancestor requirement for sites are http header [X-Frame-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options) and [Content-Security-Policy frame-ancestors directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors).
+
+However, there are application scenarios that require hosting these sites in the app's UI that is authored as an HTML page.
+`<webview>` HTML element was provided for these hosting scenarios in previous solutions like Electron and JavaScript UWP apps.
+
+For WebView2, we are providing a native API for these hosting scenarios. Developers can use it to provide additional allowed frame ancestors as if the site sent these as part of the Content-Security-Policy frame-ancestors directive. The result is that an ancestor is allowed if it is allowed by the site's original policies or by this additional allowed frame ancestors.
+  
+# Conceptual pages (How To)
+
+To host other sites in an trusted page with modified allowed frame ancestors
+- Listen to FrameNavigationStarting event of CoreWebView2.
+- Set AdditionalAllowedFrameAncestors property of the NavigationStartingEventArgs to a list additional allowed frame ancestors using the same syntax as [Content-Security-Policy frame-ancestors directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors).
+
+The list should normally only contain the origin of the top page.
+If you are hosting other sites through nested iframes and the origins of some of the intermediate iframes are different from the origin of the top page and those origins might not be allowed by the site's original policies, the list should also include those origins.
+
+You should only add an origin to the list if it is fully trusted. You should limit the usage of the API to the targetted app scenarios.
+
+# Examples
```

Comment:
Should have an outline of the algorithm here, since it's a bunch of nested callbacks.

1. (not shown) Navigate to a local page that contains
```html
<iframe src="https://example.com" name="my_site_hosting_frame"></iframe>
```
2. Wait for a frame to be created with a source equal to `https://example.com` and with a name equal to `my_site_hosting_frame`.
3. Remember that we have found the frame that is being used to hold the special content.
4. Register a listener for the frame destruction event on the frame we found in step 2.
5. When a frame navigation occurs, check if it is the frame we found in step 2.
6. If so, and the frame is navigating to `https://microsoft.com/`, then insert `https://example.com/` as an allowed frame ancestor. This allows `microsoft.com` to be loaded inside the frame.
7. When the frame from step 2 is destroyed, remember that we are no longer hosting `https://microsoft.com`.

I'm not sure I got this right.

* What triggers the frame navigation to `https://microsoft.com`?
* Why do we navigate the frame to `https://example.com/` first? Why not just use `<iframe src="https://microsoft.com" ...>`?
* In the FrameNavigating event, shouldn't we check that the frame that is navigating is the same one we found in step 2? Actually FrameNavigating doesn't tell us which frame is navigating, so we don't really know whether we are responding to the correct frame. This seems like a hole in the API.

Reply:
Also add: Note that the target site (https://microsoft.com) could also has nested iframes, you would want to add additional allowed frame ancestors for all iframes instead of just the iframe (in this example, with name as my_site_hosting_frame) directly on the top page.

Reply:
Yes, please update example code to make this more explicit.